control_plane/src/local_env.rs

```diff
@@ -325,19 +315,67 @@ pub fn create_branch(
     let oldtimelinedir = repopath
         .join("timelines")
         .join(startpoint.timelineid.to_string());
-    let mut copy_opts = fs_extra::dir::CopyOptions::new();
-    copy_opts.content_only = true;
-    fs_extra::dir::copy(
-        oldtimelinedir.join("wal"),
-        newtimelinedir.join("wal"),
-        &copy_opts,
+    copy_wal(
+        &oldtimelinedir.join("wal"),
+        &newtimelinedir.join("wal"),
+        startpoint.lsn,
+        16 * 1024 * 1024 // FIXME: assume default WAL segment size
     )?;
 
     Ok(())
 }
 
+///
+/// Copy all WAL segments from one directory to another, up to given LSN.
+///
+/// If the given LSN is in the middle of a segment, the last segment containing it
+/// is written out as .partial, and padded with zeros.
+///
+fn copy_wal(src_dir: &Path, dst_dir: &Path, upto: Lsn, wal_seg_size: u64) -> Result<()>{
+
+    let last_segno = upto.segment_number(wal_seg_size);
+    let last_segoff = upto.segment_offset(wal_seg_size);
+
+    for entry in fs::read_dir(src_dir).unwrap() {
+        if let Ok(entry) = entry {
+            let entry_name = entry.file_name();
+            let fname = entry_name.to_str().unwrap();
+
+            // Check if the filename looks like an xlog file, or a .partial file.
+            if !xlog_utils::IsXLogFileName(fname) && !xlog_utils::IsPartialXLogFileName(fname) {
```

Comment:
Seems like IsXlogFileName should accept a `&Path` or some `P: AsRef<Path>`...

Reply:
Hmm, I don't know, the argument is just a filename, not a path. For comparison, `File::file_name(&self)` also returns a string, not a path. To be precise, it returns an `&OsStr`. Maybe IsXLogFileName() should also accept `&OsStr`?

Reply:
Sure, `AsRef<OsStr>` is even more flexible. That way the argument [could be any of](https://doc.rust-lang.org/std/ffi/struct.OsStr.html#impl-AsRef%3COsStr%3E): `OsString`, `OsStr`, `String`, `str`, `Path`, `PathBuf`, etc.